### ISSUE TITLE
Updated Payments - Multiple Payment Methods

### ISSUE THREAD

**Issue by leihla** [okTurtles/group-income#1339]
## Overview

**Problem**
The current implementation only accounts for users to record manual payments that are performed out-of-platform.

 **Solution**
Add Lightning Network payments and update payments flow.

**What changed:**

- Distribution date (we now have a date where distributions are calculated and payments are expected to be done)
- Multiple payment methods (Manual + Lightning Network)

---

**Updated Flows:**

## 1. Updated - Payments with Distribution date + multiple payment methods

- [ ] Add distribution banner / chip
- [ ] Add Payment Method filters
- [ ] Add checkboxes for each payment row
- [ ] Add Accepted Methods to the "To Do" Payments Table
<img width="1440" alt="Screenshot 2022-07-09 at 11 23 28" src="https://user-images.githubusercontent.com/158178/178101703-214840fe-f663-4aa1-8233-ad88ac5ef0cf.png">

**Figma**: https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=23249%3A231771

---

## 2. New - Send Payments with Lightning Network

- [ ] Select Lighning Network Filter and check users 
- [ ] New LN Payment Modal - includes QR code for scanning payment
- [ ] New Donation fee - it's added to the transaction

<img width="1446" alt="Screenshot 2022-07-09 at 11 27 52" src="https://user-images.githubusercontent.com/158178/178101847-1d7aedaa-56be-4697-90bb-a4575f021d55.png">

**Figma**: https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=23123%3A224576

---

## 3. Updated - Record Manual Payments

- [ ] Update TODO table

<img width="1433" alt="Screenshot 2022-07-09 at 11 30 03" src="https://user-images.githubusercontent.com/158178/178104305-278224ae-74a7-4101-82c3-4cd209c6350e.png">

**Figma**: https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=23189%3A249357

---
## 4. New - Payment details Modal for Lightning

- [ ] Create Payment Details modal for LN Payments with Transaction ID

<img width="717" alt="Screenshot 2022-07-09 at 14 15 44" src="https://user-images.githubusercontent.com/158178/178107494-017c0920-9523-41ce-8f75-a249beea0e99.png">

**Figma**: https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=23123%3A225402

---
## 5. Update Receiving Payments 

- [ ] Update Receiving Payments table to add filters, Payment method and payment date

<img width="720" alt="Screenshot 2022-07-09 at 14 17 47" src="https://user-images.githubusercontent.com/158178/178107449-b1ed0cff-d2d9-41c6-b3e9-3c2fc1d6b0f2.png">

**Figma**: https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=23240%3A228725

**Comment by taoeffect** [comment]
Also related issue: https://github.com/okTurtles/group-income/issues/1240#issuecomment-1121111753

**Comment by SebinSong** [comment]
Hi @taoeffect ,

I'm currently doing some research needed for implementing sub-task 2. (**_building lightning network payment modal with dummy data_**).

<image src='https://user-images.githubusercontent.com/17641213/186370195-baaeac31-6b7e-43b7-b39d-15ce1118c6a0.png' width='420' >

as per our discussion before, I went through this[ awesome-vue](https://github.com/vuejs/awesome-vue) collection for choosing an npm pkg for QR-code generation, and decided to use [qrious](https://github.com/neocotic/qrious) as the option, as it seems pretty popular & stable. (I took time to play around with it in my [code-pen](https://codepen.io/freenaturalsoul/full/GRxeLRr) here, and it seems to work well. you can try it yourself there too)


I have a couple of questions regarding QR-code generation in the modal btw.

**Q1. What is the string that we encode within the QR-code?** if it's some sorts of a link, how do we compose that link?
One of the memos in the figma (screenshot below) says the QR-code has to be re-generated when the user takes actions (e.g. when the amount changes etc..) in the modal.

<image src='https://user-images.githubusercontent.com/17641213/186373161-513c6f79-aefe-4efc-a385-3a8030cd4d51.png' width='380'>

**Q2. is the string in the red-box the same string we bake into the QR-code?**

<image src='https://user-images.githubusercontent.com/17641213/186371442-ffc98988-e056-4dc9-b6cb-d96b254d42d0.png' width='380' >

Cheers,

**Comment by taoeffect** [comment]
@SebinSong Nice find! qrious looks great!

> Q1. What is the string that we encode within the QR-code? if it's some sorts of a link, how do we compose that link?

Unknown at this time! We're going to figure that out later. :)

> Q2. is the string in the red-box the same string we bake into the QR-code?

Although it might be, I'm guessing that it's possible that it won't be.

For example, it's possible that the string in the box will be specially encoded somehow in the QR code, for example, it could be wrapped in JSON with additional information. Unclear at this time, so please assume it's two separate, but related things.

**Comment by SebinSong** [comment]
Hi @taoeffect  @leihla ,
I'm currently working on task 5. of this issue and got a question regarding this 'NOT RECEIVED' badge component.

<img src='https://user-images.githubusercontent.com/17641213/190575932-d2df20cb-6de4-4eff-9d07-668c1108daad.png' width='480' >

In the figma, the badge doesn't display the 'NOT RECEIVED' text next to it in 'Received payments' tab(screenshot above) for all three screen sizes (mobile/tablet/desktop) but it does have the text in 'Todo' for all three screens.

<img src='https://user-images.githubusercontent.com/17641213/190578038-6c132ce1-5fcf-4ebd-b677-f365f9d4c9c3.png' width='480' > 

but how it looks currently in the app is, the badge doesn't display the text in mobile size screen but does show it in tablet/desktop screens regardless of which tab it is in.

in summary,
Is below 

[ CURRENT ]
- in **all** tabs: **hiding** the text in mobile screen , **showing** it tablet/desktop screens

-> 

[ UPDATED ]
- in 'Received payments' tab: **hiding** the texts in all mobile/tablet/desktop screens
- in 'Todo' tab: **displaying** the texts in all mobile/tablet/desktop screens

 how we want to change it?
thanks,

**Comment by leihla** [comment]
Hey @SebinSong,

**_in 'Received payments' tab: hiding the texts in all mobile/tablet/desktop screens_**
We had to add more columns to the "Received Payment" screen which made it difficult to display the entire text even on desktop, but my thoughts were that, as the not received is an action performed by the receiver (that user), they don't need to see the entire text, they just need a visual indicator of the action they performed.

**_in 'Todo' tab: displaying the texts in all mobile/tablet/desktop screens_**
On the other hand, the "Sender" needs to be notified that the payment they made was not received, and should see the entire badge as they will need to resend the payment or check what went wrong with it.

Does that make sense?

**Comment by SebinSong** [comment]
thanks, @leihla. it helps.

**Comment by SebinSong** [comment]
Hi @taoeffect,
I've got a question regarding implementing this 'Send Thank You note' feature in `group.js` contract.

<img src='https://user-images.githubusercontent.com/17641213/190844959-5e978216-288c-48b1-83ca-75d59cfb0f73.png' width='440' >

<img src='https://user-images.githubusercontent.com/17641213/190845003-993cf8de-d43c-42dd-974d-5c440ae3484d.png' width='440' >

Q. I was thinking of using `gi.contracts/group/paymentUpdate` action in `group.js` for storing this thank you note string in the corresponding payment item. to be more specific, by appending a field for 'thank you note' string to the `validate()` method of the action (below screenshot)

<img src='https://user-images.githubusercontent.com/17641213/190845240-a9202f32-1487-42b1-8ab2-f6e6603d45bb.png' width='440' >

Does this sound like a good idea? or is there somewhere else you think is more appropriate for storing it?

Thanks,

**Comment by taoeffect** [comment]
@SebinSong I think that might not be the best place for it. Reason being, we want the "thank you" to do the following:

- Trigger a bell notification for the user receiving it
- Possibly result in a DM (when DMs are implemented)
- And maybe have other side effects

And I don't think the Payment Details modal contains the 'thank you' message, right?

Therefore it should probably be a standalone action that accomplishes these things. Also, we might want it in a standalone action anyway because we might want the sideEffect to be treated in a special way in the future (I just opened #1411 for this).

**Comment by SebinSong** [comment]
@taoeffect Thanks for the answer. :)
I've got follow-up questions for that standalone action in `group.js`.

Q1. `gi.actions/group/sendPaymentThankYou` sounds good as the standalone action name?
Q2. should this standalone action be defined inside `actions` object of `frontend/models/contracts/group.js` and then be exposed as an encrypted action in `frontend/controller/actions/group.js`?
(e.g.`...encryptedAction('gi.actions/group/leaveChatRoom', L('Failed to leave chat channel.'))`)

Q3. this is a question not specific to this task but, in our code-base, some contract actions are exposed as an encrypted actions(e.g. `gi.actions/group/leaveChatRoom`) whereas the others(`gi.actions/group/create`) are directly defined in `controller/actions`. what is the difference between these two types?

Thanks,

**Comment by taoeffect** [comment]
> Q1. gi.actions/group/sendPaymentThankYou sounds good as the standalone action name?

Yep!

> Q2. should this standalone action be defined inside actions object of frontend/models/contracts/group.js and then be exposed as an encrypted action in frontend/controller/actions/group.js?

Yep!

> what is the difference between these two types?

They are essentially the same, but ones that use `encryptedAction` all do the same thing, whereas other actions will do additional things. For example, we can't use `encryptedAction` with `gi.actions/group/create`, because in addition to doing everything that `encryptedAction` does, we want it to do some other things. Look at its definition to see what those other things are.

**Comment by taoeffect** [comment]
With #1421 merged can this be closed?

**Comment by SebinSong** [comment]
@taoeffect yeah looks like it. integrating the Lightning payments would be the follow-up task of this issue. and if anything that's been missed out will be spotted in the future, I can come back and pick them up again.
